gh-111926: Make weakrefs thread-safe in free-threaded builds

mpage opened the pull request:
This makes weakrefs thread-safe in free-threaded builds. Please see the comment at the beginning of `Objects/weakrefobject.c` for an explanation of the approach.

Note that this only affects free-threaded builds. Apart from some minor refactoring, the added code is all either gated by `ifdef`s or is a no-op (e.g. `Py_BEGIN_CRITICAL_SECTION`).

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-111926 -->
* Issue: gh-111926
<!-- /gh-issue-number -->

mpage committed:
Add _PyObject_SetMaybeWeakref

We need to tag weakrefs and their referents

mpage committed:
Make PyWeakref_NewRef thread-safe

mpage committed:
Make PyWeakref_NewProxy thread-safe

mpage committed:
Make weakref___new___ thread-safe

mpage committed:
Make weakref_hash thread-safe

mpage committed:
Handle multiple weakrefs

mpage committed:
Make _PyStaticType_ClearWeakRefs thread-safe

mpage committed:
Make _PyWeakref_GetWeakrefCount thread-safe

mpage committed:
Make _weakref._getweakrefs thread-safe

mpage committed:
Make weakref clearing in subtype_dealloc thread-safe

mpage committed:
Fix incorrect size in test_sys for weakref/proxy

mpage committed:
Use a once flag

mpage committed:
Simplify clearing callback-free refs

mpage committed:
Rationalize naming

mpage committed:
Fix warning

mpage committed:
Simplify ClearWeakRefs

mpage committed:
Refactor clearing code

mpage committed:
Protect wr_object with a separate mutex

mpage committed:
Switch to a sharded lock for linked list and object

mpage committed:
Fix formatting

mpage committed:
Remove vestigial include

mpage committed:
Make sure we set the cleared flag

mpage committed:
Use `Py_TryIncref` when returning the ref

The `is_dead` check from the default build is insufficient, since the
refcount may go to zero between when we check and when we incref. More
generally, `Py_NewRef` is unsafe to use if you have a borrowed refernce.

mpage committed:
Don't leak ref if initialization fails

mpage committed:
Remove clinic critical section directives for `_weakref` module

Locking is handled in the implementation

mpage committed:
Support async removal of dead weakrefs from WeakValueDictionary

mpage committed:
Only use the striped lock

mpage committed:
Use QSBR to avoid locking in PyType_IsSubtype

mpage committed:
Refactor weakref creation

mpage committed:
Fixups

mpage committed:
Fix formatting

mpage committed:
Remove unnecessary include

mpage committed:
Fix unused function warning in default builds

colesbury commented on Include/internal/pycore_interp.h:
```diff
@@ -59,6 +59,12 @@ struct _stoptheworld_state {
+// This should be prime but otherwise the choice is arbitrary. A larger value
+// increases concurrency at the expense of memory.
+#define NUM_WEAKREF_LIST_LOCKS 127
+#endif
```
The style I've seen recommended in my PRs is to indent the `#define` here:

```suggestion
#ifdef Py_GIL_DISABLED
// This should be prime but otherwise the choice is arbitrary. A larger value
// increases concurrency at the expense of memory.
#  define NUM_WEAKREF_LIST_LOCKS 127
#endif
```

colesbury commented on Include/internal/pycore_weakref.h:
```diff
@@ -30,53 +48,82 @@ static inline int _is_dead(PyObject *obj)
+// must be held around calls to this function.
 extern Py_ssize_t _PyWeakref_GetWeakrefCount(PyWeakReference *head);
 
+extern Py_ssize_t _PyWeakref_GetWeakrefCountThreadsafe(PyObject *obj);
```
Can we combine these two functions? Just have `_PyWeakref_GetWeakrefCount(PyObject *obj)` and make it thread-safe.

mpage replied:
That would certainly make the most sense! I'm not really sure what's acceptable with respect to changing the function signature for existing functions. I guess that really only matters for exported functions?

colesbury commented on Python/pystate.c:
```diff
@@ -617,6 +625,9 @@ init_interpreter(PyInterpreterState *interp,
     _Py_brc_init_state(interp);
+    for (int i = 0; i < NUM_WEAKREF_LIST_LOCKS; i++) {
+        interp->weakref_locks[i] = (PyMutex){0};
+    }
```
We don't need to explicitly initialize these. `PyInterpreterState` is zero initialized and we don't generally explicitly initialize fields to zero.

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -48,6 +128,59 @@ new_weakref(PyObject *ob, PyObject *callback)
+            /* If 'self' is the end of the list (and thus self->wr_next ==
+               NULL) then the weakref list itself (and thus the value of *list)
+               will end up being set to NULL. */
+            _Py_atomic_store_ptr(list, self->wr_next);
```
Why does this use an atomic operation? Isn't `wr_next` protected by the striped locks?

mpage replied:
We read `wr_next` without holding the lock at the beginning of `PyObject_ClearWeakRefs` to provide a fast path for the common case where an object doesn't have any weakrefs. Since `PyObject_ClearWeakRefs` is called as part of the deallocation path for any object that *could* have weakrefs I think it makes sense to have a fast path.

colesbury replied:
Ok, that makes sense to me.

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -48,6 +128,59 @@ new_weakref(PyObject *ob, PyObject *callback)
+clear_weakref_lock_held(PyWeakReference *self, PyObject **callback)
+{
+    // TODO: Assert locks are held or world is stopped
+    if (_Py_atomic_load_ptr(&self->wr_object) != Py_None) {
```
Does this need to be atomic? Don't we only modify it with the lock held?

mpage replied:
> Don't we only modify it with the lock held?

We do, but we read it without holding the lock (see my reply above).

colesbury replied:
The reads that don't hold the lock should continue to use `_Py_atomic_load_ptr`, but the functions that hold the lock should use plain, non-atomic reads. We'll get better, more useful TSAN coverage if we don't use atomic loads unnecessarily.

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -1,10 +1,73 @@
 
+#ifdef Py_GIL_DISABLED
+/*
+ * Thread-safety for free-threaded builds
```
I think there is more background and detail here than is helpful. In the spirit of "[economy of expression](https://devguide.python.org/documentation/style-guide/#economy-of-expression)" let's trim this to the essential point:  the weakref linked list and `wr_object` are protected by a striped lock keyed by `wr_object`'s address.

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -48,6 +128,59 @@ new_weakref(PyObject *ob, PyObject *callback)
+
+// Clear the weakref while the world is stopped
+static void
+gc_clear_weakref(PyWeakReference *self)
```
Can these functions (`clear_weakref_lock_held`, `gc_clear_weakref`, `clear_weakref`) be combined with their default build implementations?

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -276,64 +426,157 @@ insert_head(PyWeakReference *newref, PyWeakReference **list)
+    PyWeakReference **list = GET_WEAKREFS_LISTPTR(obj);
+    PyWeakReference *ref, *proxy;
+
+    LOCK_WEAKREFS(obj);
```
This code structure seems more complicated than necessary because we lock twice. We only re-use weakreferences if they have no callback and one of three concrete types. Those three concrete types should be safe to allocate while holding the lock, even though it's not safe in general to allocate arbitrary types while holding the lock.

Let's restructure this so that we don't need the double checking. Something like:

1) First, check the type and callback to determine if we can reuse a weakref instance
2) If so, the with the lock held try to reuse an instance or allocate and insert a new one. Importantly the allocation and insertion happens without releasing the lock so we don't need to double check it.
3) Otherwise, allocate and return a new instance without acquiring the lock

mpage replied:
> Those three concrete types should be safe to allocate while holding the lock

I'm not sure that's true. `type` can be a subtype of `weakref`. Is it possible for subtypes to override `tp_alloc` with something that wouldn't be safe to call with the lock held?

colesbury replied:
We don't reuse instances if `type` is a subtype of `weakref`. We only consider reuse if the type is *exactly* `_PyWeakref_RefType`, `_PyWeakref_ProxyType`, or `_PyWeakref_CallableProxyType`:

https://github.com/python/cpython/blob/29829b58a8328a7c2ccacaa74c1d7d120a5e5ca5/Objects/weakrefobject.c#L240-L246

If `type` is a subtype (i.e., not one of those three exact types), then we should allocate first (without the lock) and then insert unconditionally (with the lock).

mpage replied:
Oh yeah, of course. Sorry I failed to fully comprehend your original comment.

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -276,64 +426,157 @@ insert_head(PyWeakReference *newref, PyWeakReference **list)
+    }
+    if (type == &_PyWeakref_ProxyType) {
+        cand = proxy;
+    }
```
Doesn't this also need to handle `_PyWeakref_CallableProxyType`?

mpage replied:
Yep, good catch!

colesbury commented on Include/object.h:
```diff
@@ -766,6 +766,7 @@ PyAPI_FUNC(void) _Py_NegativeRefcount(const char *filename, int lineno,
                                       PyObject *op);
 PyAPI_FUNC(void) _Py_INCREF_IncRefTotal(void);
 PyAPI_FUNC(void) _Py_DECREF_DecRefTotal(void);
+PyAPI_FUNC(void) _Py_IncRefTotal(PyInterpreterState *);
```
I don't think we want to move this

mpage replied:
This is necessary because `_Py_TryIncref` ends up using it in debug builds and `_Py_TryIncref` gets inlined into `_PyWeakref_GET_REF`, which in turn gets inlined into the sqlite and ssl modules.

An alternative would be to update those modules to use `PyWeakref_GetRef`. I think that might also let us get rid of the export for `_PyMutex_LockTimed`.

colesbury commented on Modules/_weakref.c:
```diff
@@ -86,26 +82,57 @@ Return a list of all weak reference objects pointing to 'object'.
         return NULL;
     }
 
+#ifdef Py_GIL_DISABLED
```
This seems more complicated than necessary. Don't bother with the count, just call `PyList_Append` within the loop with the lock held.

Both the default build and the free-threaded build need to handle weakrefs with zero refcount that are not yet removed from the list. Let's add a variant of `_Py_TryIncref` that works with both builds and use it here. Something like:

```
_Py_TryIncref(PyObject *op)
{
#ifdef Py_GIL_DISABLED
 ...
#else
if (Py_REFNT(op) > 0) {
    Py_INCREF(op);
    return 1;
}
return 0;
}
```

In other words, let's have one loop for both builds and avoid the list trimming logic.

mpage replied:
I *think* that's safe... my concern is that we aren't using a critical section to lock the weakref lock, so we potentially risk deadlock by doing so.

mpage committed:
Check for callable proxy

mpage committed:
Refactor weakref creation

mpage committed:
Simplify comment

mpage committed:
Get rid of GetWeakrefCountThreadsafe

mpage committed:
Don't explicitly zero initialize striped lock

mpage committed:
Indent conditional define

mpage committed:
Use non-atomic loads while we hold the lock

mpage committed:
Refactor a _PyWeakref_GET_REF and _PyWeakref_IS_DEAD to share more code

mpage committed:
Merge helper functions for clearing weakrefs

mpage committed:
Make _PyStaticType_ClearWeakRefs have one implementation for default and free-threaded builds

mpage committed:
Make sure we're able to index into the striped lock even if GC clears us

mpage committed:
Consolidate implementations of _PyWeakref_ClearRef and gc_clear

mpage committed:
Rename `_Py_TryIncref` to `_Py_TryIncrefCompare`

mpage committed:
Add `_Py_TryIncref`

mpage committed:
Refactor `weakref._getweakrefs` to have one implementation

mpage committed:
Refactor to have one implementation for `PyObject_ClearWeakRefs`

mpage committed:
Revert "Use QSBR to avoid locking in PyType_IsSubtype"

This reverts commit dd922b1f3b72e92c76942a3da1d6e486e11860e5.

mpage committed:
Accept thread unsafety when checking subtype for now

mpage committed:
Update modules to use PyWeakref_GetRef

So we don't need to export PyMutex_LockTimed and _Py_IncRefTotal

mpage committed:
Merge branch 'main' into gh-111926-thread-safe-weakref

mpage committed:
Use a pointer to the appropriate mutex rather than a copy of the object ptr

mpage committed:
Fix formatting

mpage committed:
Merge implementation of try_reuse_basic_ref

colesbury commented on Modules/_sqlite/blob.c:
```diff
@@ -102,8 +101,8 @@ pysqlite_close_all_blobs(pysqlite_Connection *self)
         PyObject *weakref = PyList_GET_ITEM(self->blobs, i);
-        PyObject *blob = _PyWeakref_GET_REF(weakref);
-        if (blob == NULL) {
+        PyObject *blob;
+        if (!PyWeakref_GetRef(weakref, &blob)) {
```
Are these changes necessary? If not, let's keep them as is.

mpage replied:
The `_sqlite` and `_ssl` modules both use some combination of `_PyWeakref_GET_REF` and `_PyWeakref_IS_DEAD`. Those are both static inline functions that end up calling unexported functions (`_PyMutex_LockTimed` and `_Py_IncRefTotal`), so linking fails without any change.

Exporting the symbols is an alternative. I don't love it, but I guess it's OK. 

Another alternative would be to add something like `_PyWeakref_IsDead` that just proxies to `_PyWeakref_IS_DEAD` and update the appropriate callsites. That would solve the symbol visibility issues and avoid changing the semantics of the calling code.

colesbury replied:
Ok, in that case I think the `_PyWeakref_GET_REF` -> `PyWeakref_GetRef` is fine.

For `_PyWeakref_IS_DEAD`, I think either:

1) Remove the locking in `_PyWeakref_IS_DEAD`. We don't really it. The call is inherently racy: just because it's not dead when you call `_PyWeakref_IS_DEAD` doesn't mean it's still alive in the next instant.
2) Or add the exported `_PyWeakref_IsDead` function (defined in `pycore_weakref.h`) for `_sqlite/connection.c`

colesbury commented on Modules/_sqlite/connection.c:
```diff
@@ -1065,9 +1064,11 @@ static void _pysqlite_drop_unused_cursor_references(pysqlite_Connection* self)
+        if (!PyWeakref_GetRef(weakref, &obj)) {
             continue;
         }
+        Py_DECREF(obj);
```
I don't love this rewrite to use `PyWeakref_GetRef` because now you have introduced a `Py_DECREF()` that possibly can call arbitrary code via a destructor.

colesbury commented on Modules/_weakref.c:
```diff
@@ -86,26 +79,39 @@ Return a list of all weak reference objects pointing to 'object'.
-    PyWeakReference *current = *list;
-    for (Py_ssize_t i = 0; i < count; ++i) {
-        PyList_SET_ITEM(result, i, Py_NewRef(current));
+    LOCK_WEAKREFS(object);
+    PyWeakReference *current = *GET_WEAKREFS_LISTPTR(object);
+    while (current != NULL) {
+        PyObject *curobj = (PyObject *) current;
+        if (_Py_TryIncref(curobj)) {
+            if (PyList_Append(result, curobj)) {
+                UNLOCK_WEAKREFS(object);
+                Py_DECREF(curobj);
+                Py_DECREF(result);
+                return NULL;
+            }
+            else {
+                // Undo our _Py_TryIncref. This is safe to do with the lock
+                // held in free-threaded builds; the list holds a reference to
+                // curobj so we're guaranteed not to invoke the destructor.
+                Py_DECREF(curobj);
+            }
+        }
         current = current->wr_next;
     }
+    UNLOCK_WEAKREFS(object);
```
Nice. I think this fixes some bugs with `_weakref.getweakrefs` returning dead references.

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -49,25 +88,19 @@ new_weakref(PyObject *ob, PyObject *callback)
     if (self->wr_object != Py_None) {
         PyWeakReference **list = GET_WEAKREFS_LISTPTR(self->wr_object);
-
         if (*list == self)
```
There are enough changes to this function that I think adding the missing `{` `}` to if statements makes sense.

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -942,68 +1022,73 @@ PyObject_ClearWeakRefs(PyObject *object)
+        return;
+    }
+
+    int num_items = 0;
```
Let's use `Py_ssize_t` for sizes instead of `int`

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -942,68 +1022,73 @@ PyObject_ClearWeakRefs(PyObject *object)
+        Py_XDECREF(callback);
     }
+
+    for (int i = 0; i < num_items; i += 2) {
```
Same here: `Py_ssize_t`

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -942,68 +1022,73 @@ PyObject_ClearWeakRefs(PyObject *object)
-            }
+        done = (*list == NULL) || !is_basic_ref_or_proxy(*list);
+        UNLOCK_WEAKREFS(object);
+        Py_XDECREF(callback);
```
`callback` must be `NULL` here because we only clear basic refs. Let's:

1) Move the `PyObject *callback` definition inside the if statements scope
2) Assert that it's `NULL` instead of decrefing it

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -1017,10 +1102,29 @@ _PyStaticType_ClearWeakRefs(PyInterpreterState *interp, PyTypeObject *type)
+        done = (*list == NULL);
+        UNLOCK_WEAKREFS(type);
+        Py_XDECREF(callback);
+    }
```
This is called at shutdown when there is only one thread left. We don't need to do any locking. I think the following code should be sufficient:

```
    PyObject **list = _PyStaticType_GET_WEAKREFS_LISTPTR(state);
    while (*list) {
        _PyWeakref_ClearRef((PyWeakReference *)*list);
    }
```

mpage replied:
This slightly changes the semantics - the original version decrefs the callback. Doing it this way postpones that decref until the weakref's destructor is invoked, which is probably fine.

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -776,104 +920,40 @@ _PyWeakref_CallableProxyType = {
-        Py_INCREF(result);
+    if (PyCallable_Check(obj)) {
+        Py_SET_TYPE(result, &_PyWeakref_CallableProxyType);
+    }
```
This is icky. Can we refactor this so that:

1) The type determination is pulled up into `PyWeakref_NewProxy`
2) `new_weakref` takes the type so that it doesn't have to be changed after creation

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -276,64 +330,154 @@ insert_head(PyWeakReference *newref, PyWeakReference **list)
+    PyWeakReference **list = GET_WEAKREFS_LISTPTR(obj);
+    if ((type == &_PyWeakref_RefType) ||
+        (type == &_PyWeakref_ProxyType) ||
+        (type == &_PyWeakref_CallableProxyType)) {
```
Put the `{` on it's own line here (see [PEP 7](https://peps.python.org/pep-0007/#:~:text=When%20you%20break%20a%20long%20expression%20at%20a%20binary%20operator%2C%20the%20operator%20goes%20at%20the%20end%20of%20the%20previous%20line%2C%20and%20braces%20should%20be%20formatted%20as%20shown.%20E.g.%3A))

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -49,25 +88,19 @@ new_weakref(PyObject *ob, PyObject *callback)
 {
-    PyObject *callback = self->wr_callback;
-
+    // TODO: Assert locks are held or world is stopped
```
Let's either add the assertion or remove the TODO

colesbury commented on Objects/weakrefobject.c:
```diff
@@ -276,64 +330,154 @@ insert_head(PyWeakReference *newref, PyWeakReference **list)
+                                             PyObject *);
+
+static PyWeakReference *
+get_or_create_weakref(PyTypeObject *type, weakref_alloc_fn allocate,
```
I think the `weakref_alloc_fn allocate` complicates this and is avoidable:

1) The generic allocation via `type->tp_alloc` always works
2) If you want, you can use the `new_weakref()` call with the `if` and the `type->tp_alloc` code path in the `else` (`new_weakref()` is possibly a bit faster)

colesbury commented on Include/cpython/weakrefobject.h:
```diff
@@ -30,6 +30,14 @@ struct _PyWeakReference {
+     * Normally this can be derived from wr_object, but in some cases we need
+     * to lock after wr_object has been set to Py_None.
+     */
+    struct _PyMutex *weakrefs_lock;
```
I suspect we could get rid of this with some effort, but maybe that's not worth it. In particular, we probably want this pattern if we get rid of the striped locks in the future.

mpage replied:
Yeah, I have an idea for how we could get rid of this but the correctness of the resulting code is subtle. This is easier to reason about and as you said, we're probably going to want something like this when we get rid of striped locks in the future.

colesbury reviewed (commented):
This is looking good. Some comments inline.

Additionally, `proxy_dealloc` should remove the redundant `PyObject_GC_UnTrack` inside the if statement. It's unnecessary and the plain read of `self->wr_callback` is possibly a data race.

mpage committed:
Remove redundant gc untrack

mpage committed:
Update comment

mpage committed:
Add missing braces

mpage committed:
Use Py_ssize_t for tracking number of items

mpage committed:
Tighten up clearing basic refs/proxies

mpage committed:
Simplify _PyStaticType_ClearWeakRefs

mpage committed:
Simplify get_or_create_weakref

mpage committed:
Add _PyWeakref_IsDead

mpage committed:
Remove TODO

We cannot assert now that we clear weakrefs without holding the
lock in _PyStaticType_ClearWeakRefs

mpage committed:
Fix formatting

colesbury reviewed (approved):
LGTM

erlend-aasland commented on Python/pystate.c:
```diff
@@ -506,6 +506,14 @@ _PyRuntimeState_ReInitThreads(_PyRuntimeState *runtime)
     }
+#ifdef Py_GIL_DISABLED
+    for (PyInterpreterState *interp = runtime->interpreters.head;
+         interp != NULL; interp = interp->next) {
```
```suggestion
         interp != NULL; interp = interp->next)
    {
```

colesbury committed:
Update Python/pystate.c

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>
